### DISCUSSION TITLE
Using Rules.mk for modifying tables

### DISCUSSION THREAD

**Discussion by arco400** [IBM/ibmi-tobi#309] [General]
Our environment uses almost 100% SQL source based tables. A typical source definition would contain something like:

`CREATE OR REPLACE TABLE My_Transaction (`
`  Transaction_ID FOR TRANSID INTEGER GENERATED ALWAYS AS IDENTITY`
`   (START WITH 1 INCREMENT BY 1 NO MINVALUE NO MAXVALUE NO CYCLE ORDER NO CACHE ),`
`  -- other colums`
`  Creation_TimeStamp FOR CRTTIME TIMESTAMP NOT NULL DEFAULT CURRENT TIMESTAMP,`
`  Update_TimeStamp FOR UPDTIME TIMESTAMP NOT NULL FOR EACH ROW ON UPDATE AS ROW CHANGE TIMESTAMP,   `
`  CONSTRAINT My_Long_Named_Table_Key_1 PRIMARY KEY(Transaction_ID)`
`  )`
`RCDFMT MYTRANSRF ;`

`RENAME TABLE My_Transaction TO SYSTEM NAME MYTRANS ;`

When a column needs to be added, I need to preserve the existing data and timestamps. To perform that task in our current build method, I currently modify a custom CL build script to do an "ALTER TABLE ADD COLUMN" instead of replacing the table.

Would it be possible to have such functionality in the Rules.mk definition? Like the PHONY functionality of a makefile: 

`.PHONY: all backup alter update_constraints test`

`all: backup alter update_constraints test`

`backup:`
`    # Perform database backup here`

`alter:`
`    # Execute SQL ALTER TABLE statement to add the new column`

`update_constraints:`
`    # Update constraints and foreign keys if necessary`

`test:`
`    # Run tests to ensure everything works as expected`

Thanks.

Best regards,
Arco Simonse

**Comment by edmundreinhardt** [comment]
@arco400 
Would it be possible to put that backup/alter/restore SQL statements in the .table file?

**Reply by arco400** [reply]
@edmundreinhardt Sure those ALTER and other statements can be placed in a .table source file. But I also need to update the "base" source file for the table, so that a full build will use that source.
So the thing is: How do I prevent execution of the "base" source when doing an incremental build, and how do I prevent execution of the "alter" source when doing a full build.

**Reply by edmundreinhardt** [reply]
I know it is not optimal, but could the base be:
```
CREATE TABLE xxx (
...
)

ALTER TABLE ADD COLUMN ...
```
so it would work for both create and update?

Alternatively, we could have parameter if the file exists
and the .table could have a
```
IF table_exists
   CREATE TABLE xxx (
   ...
   )
ELSE
   ALTER TABLE ...
END IF
```
this would have to be a significant enhancement.  We would have to create an SQL procedure and compile it every time.  Or we do our own preprocessing of that file and execute the SQL statements one by one.

**Reply by arco400** [reply]
The first option to do the original create step first, and alter later in the source is not optimal. But will BOB continue with the alter when the first create statement has failed (because the table already exists) and I suppose this will always report failed build items? It seems to me this is not workable for cd/ci.

That is why I asked about having something working similar to .PHONY.
If we could add a parameter with makei that can be used to indicate that it is full build or incremental, and use that in rules.mk to decide which source to execute.

**Reply by edmundreinhardt** [reply]
So is the prior existence of the file not sufficient to determine if it is a full build or incremental?
`makei build` can be used for both full and incremental depending on what exists in the target library.

**Comment by arco400** [comment]
Yes that existence would be sufficient. Maybe we don't get each other right. My only goal is to safely add a column, with preserving all timestamps. And I would be able to do full builds for testing, and incremental builds for deployments on existing environments.

**Reply by edmundreinhardt** [reply]
I think we are understanding each other.
So we are now moving to solution 2 which is to have both the complete and the incremental instructions in the source for the table and to conditionally select the appropriate one.

Some of the problems are 
1) how many changes should be kept in the incremental.  For example, I add one column this week and another column next week.  Do we only keep the latest alter in the incremental section and assume that all tables have been updated for the previous change>

2) how to do you want to do the backup and restore.  can this be done generically by BOB or should the SQL statements to do that be part of the incremental instructions

so perhaps the .table file looks like
```
incremental {
 ALTER TABLE ADD COLUMN ...
 }
 CREATE OR REPLACE TABLE ...
```

@irfanshar what i am thinking is that we try to detect the existence of the target  in the extractandlaunch.sh for SQL and if it existsn then we only extract the incremental SQL statements and pass to launch, and otherwsie we extract the remainder and add to launch.

@arco400 could you do your backup and restore instructions in the incremental section?

**Reply by irfanshar** [reply]
@edmundreinhardt Are we checking the objlib, curlib and the library list for the existence of the target?

**Reply by edmundreinhardt** [reply]
Yes we are, by using VPATH to have the directories corresponding to the library list in it

**Comment by arco400** [comment]
The first issue regarding multiple possible alters is the reason that I keep the create and alter sources separate in my current company CL based build process.
When I extend the table, I modify the source and touch it with an older date so that it not will be used for incremental builds. I put the alters per modification project in a separate source. When doing a complete build I then ignore the modification sources.

There is also an issue in that you can check for the table existence, but do not know which alters already have been applied.
For that you would need to parse the table information of the existing table I guess.

Regarding backup, if the table is not extended by a "replace" but with an "alter table add column" is done this would be gracefully handled by sql itself.

**Reply by edmundreinhardt** [reply]
So in my proposed solution,  there is only 1 source, no needing to mess with touch.
When you make a change to the table, you make it to the CREATE TABLE (not need for OR REPLACE any more).
At the same time you replace the incremental section with the `ALTER`s necessary to update an existing table.

For old table that need multiple incremental sections these can be retrieved from git.  You might even have the convention of leaving them in the table commented out.  How likely is this scenario?

**Reply by arco400** [reply]
@edmundreinhardt I believe your proposed solution might work well. That would be great if it could be implemented.

**Comment by rbeckham** [comment]
@arco400 Am i missing something?  Why could you not just change your main table statement to a "CREATE or REPLACE TABLE...".
Add the new column to this statement.  Then the next time you build your application, if the table already exists, then the data is preserved and the new column is added automatically.
Why do you need to add the complexity.....just wondering.  

I did a quick test and created a test table using the CREATE OR REPLACE TABLE with the source you gave in the comments,  I added data and timestamps were updated accordingly. I went back and added a new column to the prior CREATE oR REPLACE TABLE command and the new column was added and the data and timestamps were preserved.  Would that not be ideal?  The default settings of the CREATE OR REPLACE TABLE is to preserve the data if the table already exists.

**Comment by arco400** [comment]
@rbeckham Whow, thank you!! I have completely missed that standard functionality. Using ALTERs for so many (20+) years and I have always been under impression that a REPLACE did not preserve the existing data. 

@edmundreinhardt My apologies that I have put this question on your desk. It is unnecessary, the standard CREATE OR REPLACE TABLE fulfills this already.

**Comment by rbeckham** [comment]
@arco400  Awesome!! Glad to help!  I know that will simplify things a lot when you are just wanting to add a column or make a basic change to a table.

I don't think @edmundreinhardt work will be in vain though, since there may be more complicated scenarios where your initial request will come in handy, like when changing a column from char to numeric where some one time manual casting of data might have to be done.

I have seen this done before when I work in  Android Development Studio when you have a local sql database as part of your application.
Every point in the life of the database has a version number.  So your first build, the database would be version model 1.  When you want to change the database you have to increment to version model 2, and in your setup files you tell the application what sql has to run to upgrade your database from version 1 to version 2.  And then it just goes down the recursive tree.  All the code stays in your build files so its very possible to go from any version to the latest version and your application will still build correctly and have the right database model.

So if its a brand new build on a server you would automatically get the latest database model, whatever the latest version is.  But lets say you  are building to library where the database model is on version2. and the latest is 15......then the build knows how to get there.

Apple and Swift basically does it the same way as well...

**Comment by AIforIBMi** [comment]
@irfanshar Irfan
In the documentation, add a version comment to the first line of the pseudo source (if we support comments)
@arco400 we have it coded in a branch as you can see in the pull requests.  We are just in need of some documentation and examples.
@rbeckham Would you have an example we can put in the docs?
